src/cmake-tools.ts

```diff
@@ -271,8 +271,15 @@ export class CMakeTools implements vscode.Disposable, api.CMakeToolsAPI {
           const cmakeListsFile = await vscode.window.showOpenDialog(openOpts);
           if (cmakeListsFile) {
             const fullPathDir: string = path.parse(cmakeListsFile[0].fsPath).dir;
-            const relPathDir: string = path.relative(this.folder.uri.fsPath, fullPathDir);
-            vscode.workspace.getConfiguration('cmake').update("sourceDirectory", path.join("${workspaceFolder}", relPathDir));
+            const relPathDir: string = lightNormalizePath(path.relative(this.folder.uri.fsPath, fullPathDir));
+            const drv = await this.getCMakeDriverInstance();
+            if (drv) {
+              const joinedPath = "${workspaceFolder}/".concat(relPathDir);
+              vscode.workspace.getConfiguration('cmake', this.folder.uri).update("sourceDirectory", joinedPath);
+              drv.config.updatePartial({sourceDirectory: joinedPath});
+            } else {
+              throw new Error(localize('unable.to.save.sourceDirectory', 'Unable to save "cmake.sourceDirectory". There is no valid cmake driver instance.'));
```

Comment:
You don't need a `drv` to set `cmake.sourceDirectory`, so I think you should set it unconditionally. In that case you could just instruct the user to reload the window and the setting should take effect.  E.g. "Something went wrong while setting the cmake.sourceDirectory setting. Please run the "Reload Window" command for the change to take effect") Or give them a button to run the command automatically.